### DISCUSSION TITLE
mdast-util-from-markdown - leading and trailing whitespace inside "strong" or "italic" nodes

### DISCUSSION THREAD

**Discussion by gschlager** [syntax-tree/unist#60] [Q&A] [answered]
I noticed that `mdast-util-from-markdown` outputs invalid Markdown when a `strong` or `italic` node contains leading and/or trailing whitespace. I can implement a workaround in my code that creates the mdast, but I wonder if this should be handled in the util instead. What do you think?

Here's a test case:

``` js
import { toMarkdown } from "mdast-util-to-markdown";

const tree = {
  type: "root",
  children: [
    {
      type: "paragraph",
      children: [
        { type: "text", value: "foo" },
        {
          type: "strong",
          children: [{ type: "text", value: " this is bold " }],
        },
        { type: "text", value: "bar" },
      ],
    },
  ],
};

console.log(toMarkdown(tree));
```

**Current output:** `foo** this is bold **bar`

**Expected output:** `foo **this is bold** bar`

**Comment by ChristianMurphy** [comment]
An interesting question @gschlager.
Could you expand on how the mdast is generated?

The content example in mdast shared itself is not valid markdown.

```js
const tree = {
  type: "root",
  children: [
    {
      type: "paragraph",
      children: [
        { type: "text", value: "foo" },
        {
          type: "strong",
          children: [{ type: "text", value: " this is bold " }],
        },
        { type: "text", value: "bar" },
      ],
    },
  ],
};
```

adds spaces around `this is bold ` where spaces cannot exists in valid markdown.
`mdast-util-from-markdown` would need to change the content from the mdast (adding and removing spaces) to produce bold.

I'm not sure I see the stringifier as being the place to put content validation/content fixing.
It would be best if the process generating the mdast produces valid mdast.
If it can't the next best step would be to add in https://github.com/syntax-tree/mdast-util-assert to see if the tree is valid (adding new validations if needed)
A fixer would likely be it's own thing, and only applied where needed. It would potentially be fairly specific to how an individual project wants to rework invalid content.

**Reply by gschlager** [reply]
Thanks for the quick response!

Yes, I'm aware that it's not valid Markdown. That's why I'm wondering if it makes sense to fix it in `mdast-util-from-markdown` before outputting the Markdown. I see that the util applies other fixes as well in order to prevent it from producing invalid Markdown.

I'm implementing a converter from BBCode to Markdown, so it's definitely possible to fix it in my code before creating the mdast. However, the problem exists for HTML as well.

```html
<!doctype html>
<head><title>Test</title></head>
<p>foo<strong> this is bold </strong>bar</p>
```

`hast-util-to-mdast` converts this into `foo** this is bold **bar` as well.

**Reply by ChristianMurphy** [reply]
🤔 Thoughts @wooorm or others on if it would make sense to make changes to `hast-util-to-mdast` vs `mdast-util-to-markdown`?
I'd lean a bit toward, `hast-util-to-mdast` myself, it feels like these cases mostly happen when translating between languages, and how to handle it may also be language specific.
Though I'm open to a `mdast-util-to-markdown` solution as well.

**Reply by wooorm** [reply]
Hi there!

From the perspective of `mdast-util-to-markdown`, I think garbage in garbage out seems reasonable here.

> I see that the util applies other fixes as well in order to prevent it from producing invalid Markdown.

Which ones do you mean?

> `hast-util-to-mdast`

In the case of `hast-util-to-mdast`, such markup is not garbage, so I can see it being solved there. I agree with @ChristianMurphy.

>  I can implement a workaround in my code that creates the mdast, […]

I think this case is similar to `hast-util-to-mdast`. If it is sensible markup in BBCode, it should map to sensible markup in markdown. And the map function is resonsible there.

**Reply by gschlager** [reply]
> > I see that the util applies other fixes as well in order to prevent it from producing invalid Markdown.
> 
> Which ones do you mean?

I was thinking about the escaping of unsafe characters which have special meaning in Markdown. Or [this](https://github.com/syntax-tree/mdast-util-to-markdown/blob/52c18b4c59cd8e1580a6b837b22a8fa890fe7654/lib/handle/list.js#L44-L69) handling of a special case inside lists. Or the escaping and padding with spaces in [inline-code](https://github.com/syntax-tree/mdast-util-to-markdown/blob/main/lib/handle/inline-code.js).

The existing code tries very hard to output valid Markdown. That makes a lot of sense to me. I don't have to know the Markdown spec in order to create a `mdast`. As long as it's a valid tree, the util should take care of outputting correct Markdown.

I'd argue that handling leading/trailing whitespaces in `strong` and `emphasis` falls into the same category as the other "fixes" I mentioned.

> I think this case is similar to `hast-util-to-mdast`. If it is sensible markup in BBCode, it should map to sensible markup in markdown. And the map function is resonsible there.

Sure, I can definitely implement it in my mapping function if the decision is that this doesn't belong into `mdast-util-to-markdown`.

**Reply by wooorm** [reply]
All those cases you list solve *valid* markdown though. Things can be escaped in markdown. Lists can exist in lists. And code can be padded. Leading/trailing whitespace in emphasis, strong, and strikethrough is not valid markdown.

**Reply by gschlager** [reply]
I fully understand your point. I wasn't thinking about it in the context of valid/invalid Markdown. For me it's more about transforming text based on where it's used.

* `{ type: "text", value: " this is bold " }` is a valid text node
* as a child of "strong" it needs to be outputted differently

In my mind that's similar to the escaping that's already implemented.
* `{ type: "text", value: "# this is not a heading" }` is a valid text node
* as a child of a "strong" it can be outputted as-is
* as a child of a "paragraph" the "#" needs escaping

That's why I thought it would make sense to implement it in `mdast-util-to-markdown`, so that everyone could profit from it. Anyway, I'm happy to implement it myself in my converter.

**Reply by wooorm** [reply]
new thread: https://github.com/syntax-tree/unist/discussions/60#discussioncomment-2111096

**Comment by wooorm** [comment, marked as answer]
Hey again @gschlager! Sooo, it took me hours and hours of thinking but I managed to reproduce your AST in markdown 😅

```
fo&#x6f;**&#x20;this is bold&#x20;**&#x62;ar
```
->
fo&#x6f;**&#x20;this is bold&#x20;**&#x62;ar

Pretty doubtful that anyone would ever write that though. But even before I figured it out I already become more understanding of the use cases you mention, where folks are working on the ASTs and injecting punctuation/whitespace/whatever in text or adding/removing emphasis nodes.

We already use character references in a couple places. When things can be input:

```markdown
&#x20;this&#x20;
```
->
```html
<p> this </p>
```

…we try and output it too:

```js
import {toMarkdown} from 'mdast-util-to-markdown'

/** @type {import('mdast').Root} */
const tree = {
  type: 'root',
  children: [
    {type: 'paragraph', children: [{type: 'text', value: ' this '}]}
  ]
}

console.log(toMarkdown(tree))
```
->
```markdown
&#x20;this&#x20;
```

That way, it roundtrips. We can turn anything into a character reference.
We could do something like that for emphasis too. A few examples of emphasis:

```markdown
x * this *

x *&#x20;this&#x20;*

x *.this.*

x *&#x2e;this&#x2e;*
```
->
```html
<p>x * this *</p>
<p>x <em> this </em></p>
<p>x <em>.this.</em></p>
<p>x <em>.this.</em></p>
```

Emphasis can form based on the kind of character before and after the “run”, which can be whitespace, punctuation, or anything else.

A bigger example (though still reduced, because left and right runs are the same and only looking at asterisks):

```markdown
|                         | A (letter inside) | B (punctuation inside) | C (whitespace inside) | D (nothing inside) |
| ----------------------- | ------------- | ------------------ | ----------------- | -------------- |
| 1 (letter outside)      | x*y*z         | x*.*z              | x* *z             | x**z           |
| 2 (punctuation outside) | .*y*.         | .*.*.              | .* *.             | .**.           |
| 3 (whitespace outside)  | x *y* z       | x *.* z            | x * * z           | x ** z         |
| 4 (nothing outside)     | *x*           | *.*                | * *               | **             |
```
->
|                         | A (letter inside) | B (punctuation inside) | C (whitespace inside) | D (nothing inside) |
| ----------------------- | ------------- | ------------------ | ----------------- | -------------- |
| 1 (letter outside)      | x*y*z         | x*.*z              | x* *z             | x**z           |
| 2 (punctuation outside) | .*y*.         | .*.*.              | .* *.             | .**.           |
| 3 (whitespace outside)  | x *y* z       | x *.* z            | x * * z           | x ** z         |
| 4 (nothing outside)     | *x*           | *.*                | * *               | **             |

Inspecting that, we can divide them into two groups:
Things that don’t form emphasis. List X: 1B, 1C, 1D, 2C, 2D, 3C, 3D, 4C, 4D.
Things that do form emphasis. List Y: 1A, 2A, 2B, 3A, 3B, 4A, 4B.

Now given our magic trick: we can turn letters (1 and A) and whitespace (3 and C) into punctuation (2 and B), by turning them into references (which start with `&` and end in `;`). We can use that to turn entries from one group into the other.
And we want to change entries from their group to the other one.

- 1B (list X; letter outside, punctuation inside) into into 2B (list Y, punctuation inside & outside). ✅
- 1C (X) -> 2C (X) -> 2B (Y), 1B (X) -> 2B (Y) ✅
- 1D (X) -> 2D (X) 💥
- 2C (X) -> 2B (Y) ✅
- 2D (X) 💥
- 3C (X) -> 2C (X) -> 2B (Y), 3B (Y) -> 2B (Y) ✅
- 3D (X) -> 2D (X)
- 4C (X) -> 4B (Y) ✅
- 4D (X) 💥
- 1A (Y) -> 2A (Y) -> 2B (Y), 1B (X) ✅ -> 2B (Y)
- 2A (Y) -> 2B (Y) 💥
- 2B (Y) 💥
- 3A (Y) -> 3B (Y) -> 2B (Y), 2A (Y) -> 2B (Y) 💥
- 3B (Y) -> 2B (Y) 💥
- 4A (Y) -> 4B (Y) 💥
- 4B (Y) 💥

Visualizing that:

|                         | A (letter inside) | B (punctuation inside) | C (whitespace inside) | D (nothing inside) |
| ----------------------- | ------------- | ------------------ | ----------------- | -------------- |
| 1 (letter outside)      | ✅ (Y)        | ✅ (X)             | ✅ (X)           | 💥 (X)         |
| 2 (punctuation outside) | 💥 (Y)        | 💥 (Y)             | ✅ (X)            | 💥  (X)         |
| 3 (whitespace outside)  | 💥 (Y)        | 💥 (Y)             | ✅ (X)            | 💥 (X)        |
| 4 (nothing outside)     | 💥 (Y)        | 💥 (Y)             | ✅ (X)            | 💥 (X)            |

We observe some interesting aspects:

1. Whitespace inside (column C) solves everything
2. Nothing can be done for nothing inside (D)
3. The problem except for D is with list Y: things that do form emphasis, and how to change them into non-emphasis.

In this case though, we only care about list X: we’re looking at an AST `emphasis` node, which currently creates non-emphasis, and want to solve it by turning it into emphasis!
So our only problem is empty emphasis, whatever it’s surrounded by! So this:

```js
import {toMarkdown} from 'mdast-util-from-markdown'

/** @type {import('mdast').Root} */
const tree = {type: 'root', children: [
  {type: 'paragraph', children: [{type: 'text', value: 'a '},
  {type: 'emphasis', children: []},
  {type: 'text', value: ' b'}]}
]}

console.log(toMarkdown(tree))
```
-> (current)
```markdown
a ** b
```
->
a ** b

From the interesting aspects above, we found that we can add an encoded whitespace (a normal space, zero-width space? No break space?) inside it:

```markdown
punctuation around:
a.*&#x20;*.b (space)
a.*&#x200b;*.b (zwsp)
a.*&#xA0;*.b (nbsp)

whitespace around:
a *&#x20;* b (space)
a *&#x200b;* b (zwsp)
a *&#xA0;* b (nbsp)
```
->
punctuation around:
a.*&#x20;*.b (space)
a.*&#x200b;*.b (zwsp)
a.*&#xA0;*.b (nbsp)

whitespace around:
a *&#x20;* b (space)
a *&#x200b;* b (zwsp)
a *&#xA0;* b (nbsp)

It’s not perfect, adding that space, as expressed before.
Inspecting the above result though, the normal space inside the emphasis when surrounded by more whitespace, displays as nothing. Browsers ignore the empty emphasis. Perhaps we can also drop the empty emphasis?

a  b?

**Reply by gschlager** [reply]
Wow, that's a very detailed analysis. Thank you for that.

The workaround I was planning to implement would have removed the leading/trailing whitespace and put it outside of the emphasis. It results in cleaner Markdown, but I understand that you want to be able to roundtrip, so encoding some chars is probably the only solution. It's ugly, but it works, so I'd be fine with it. :slightly_smiling_face: 

> we can add an encoded whitespace (a normal space, zero-width space? No break space?)

For simplicity I'd go with a normal space.

> Perhaps we can also drop the empty emphasis?

An empty emphasis doesn't make much sense in my book. I'd drop it.

**Reply by wooorm** [reply]
I stumbled upon this again but I forgot whether this has landed already? Is this still a problem?

**Reply by wooorm** [reply]
Done in https://github.com/syntax-tree/mdast-util-to-markdown/releases/tag/2.1.1. `npm update` to get the fix! Work like this takes time. You can [sponsor us](https://unifiedjs.com/community/sponsor/) so that we can spend that time

**Comment by bakkerjoeri** [comment]
I stumbled upon this thread trying to figure out how to deal with whitespace in inline HTML elements (created in some wysiwyg editor) when converting it to markdown.

I'm running this:

```
<p>This is some<strong> sentence</strong> with some <em>inline </em>styling.</p>
```

... through `hast-util-from-html/fromHtml` -> `hast-util-to-mdast/toMdast` -> `mdast-util-to-markdown/toMarkdown`, which outputs:

```
This is some** sentence** with *inline *styling.
```

Which doesn't work as markdown:

> This is some** sentence** with *inline *styling.

Similarly if `<br />` linebreaks occur inside the `strong` or `em` tags.

So unlike @gschlager I didn't get it from parsing markdown but from parsing HTML to hast to mdast to markdown, still leading to the same invalid output. I'm trying to figure out if I can somehow "sanitize" the whitespace characters to lift them out of the inline elements to prepare for the markdown transformation but I haven't figured it out yet.

**Reply by wooorm** [reply]
a naive solution is to look through the AST, for emphasis/strong that starts/ends with whitespace, and to move that whitespace outside instead?

**Reply by bakkerjoeri** [reply]
Yeah, that's what I'm thinking if there is no existing solution, although my experience traversing and editing AST's is pretty limited so I'll have to bang my head against it for a bit. At least I'll learn some new things hehe.

**Reply by wooorm** [reply]
Alternatively you can also start looking into the to-markdown code for emphasis, and then try to solve the issue at hand, with my above info that you can turn a space into & #x20;?

**Reply by bakkerjoeri** [reply]
I have now come up with this:

```ts
export function cleanUpSpaces(htmlTree: HastNode): HastNode {
  visit(htmlTree, (node, index, parent) => {
    if (!index) {
      return;
    }

    if (node.type !== "element") {
      return;
    }

    if (node.tagName !== "strong" && node.tagName !== "em") {
      return;
    }

    if (node.children.length === 0) {
      return;
    }

    const firstChild = node.children[0];
    if (firstChild.type === "text") {
      const [, startTrim, trimmedValue] =
        firstChild.value.match(/^(\s*)([^\s].*)/) || [];

      if (startTrim && startTrim.length > 0) {
        firstChild.value = trimmedValue;
        parent?.children.splice(index, 0, {
          type: "text",
          value: startTrim,
        });
      }
    }

    const lastChild = node.children[node.children.length - 1];
    if (lastChild.type === "text") {
      const [, trimmedValue, endTrim] =
        lastChild.value.match(/([^\s]*)(\s*)$/) || [];

      if (endTrim && endTrim.length > 0) {
        lastChild.value = trimmedValue;
        parent?.children.splice(index + 1, 0, {
          type: "text",
          value: endTrim,
        });
      }
    }
  });

  return htmlTree;
}
```

which at least passes these tests:

```ts
  test("sanitizes spaces at the start of emphasis", () => {
    expect(
      toHtml(
        cleanUpSpaces(
          fromHtml("<p>A<em> word</em> with emphasis.</p>", { fragment: true })
        )
      )
    ).toBe("<p>A <em>word</em> with emphasis.</p>");
  });

  test("sanitizes spaces at the start of strong", () => {
    expect(
      toHtml(
        cleanUpSpaces(
          fromHtml("<p>A<strong> word</strong> with strength.</p>", {
            fragment: true,
          })
        )
      )
    ).toBe("<p>A <strong>word</strong> with strength.</p>");
  });

  test("sanitizes spaces at the end of emphasis", () => {
    expect(
      toHtml(
        cleanUpSpaces(
          fromHtml("<p>A <em>word </em>with emphasis.</p>", { fragment: true })
        )
      )
    ).toBe("<p>A <em>word</em> with emphasis.</p>");
  });

  test("sanitizes spaces at the end of strong", () => {
    expect(
      toHtml(
        cleanUpSpaces(
          fromHtml("<p>A <strong>word </strong>with strength.</p>", {
            fragment: true,
          })
        )
      )
    ).toBe("<p>A <strong>word</strong> with strength.</p>");
  });
```

**Reply by danburzo** [reply]
Also stumbled on this problem recently. Apparently it’s really easy to produce trailing spaces inside emphasis nodes with WYSIWYG editors. I’ve fixed it for my case at the MDAST level with a riff on @bakkerjoeri’s code:

```js
/*
	Remark plugin to fix trailing whitespace in inline nodes.
	See: https://github.com/orgs/syntax-tree/discussions/60
*/
function remarkTrailingWhitespace() {
	return function (tree, file) {
		visit(tree, ['strong', 'emphasis'], (node, index, parent) => {
			// Remove empty nodes
			if (!node.children.length) {
				parent.children.splice(index, 1);
				return SKIP;
			}

			const first = node.children[0];

			// Remove trailing whitespace at the beginning
			if (first.type === 'text') {
				const start_match = first.value.match(/^(\s+)(.*)/);
				if (start_match) {
					const [_, start_ws, start_text] = start_match;
					parent.children.splice(index, 0, {
						type: 'text',
						value: start_ws
					});
					if (start_text) {
						// has non-whitespace content
						first.value = start_text;
					} else {
						// all-whitespace content
						node.children.shift();
					}
					// Re-visit this node
					return [SKIP, index - 1];
				}
			}

			const last = node.children[node.children.length - 1];
			if (last.type === 'text') {
				const end_match = last.value.match(/(.*?)(\s+)$/);
				if (end_match) {
					const [_, end_text, end_ws] = end_match;
					parent.children.splice(index + 1, 0, {
						type: 'text',
						value: end_ws
					});
					if (end_text) {
						last.value = end_text;
					} else {
						node.children.pop();
					}
				}
			}

			return SKIP;
		});
	};
}
```

It doesn’t cover all the scenarios but was good enough to solve the problem at hand.

**Reply by thepont** [reply]
I didn't use any of the above solutions because of the edge cases (bold inside italic) etc., So I came up with this very hacky solution just adding a 0 width space to any offending text.

```javascript
import {visit, SKIP, CONTINUE} from 'unist-util-visit';

const REMOVE_TYPES = ['strong', 'emphasis'];

/*
	Remark plugin to fix trailing whitespace in inline nodes.
	See: https://github.com/orgs/syntax-tree/discussions/60
*/

const START_SPACE = /^\s/;
const END_SPACE = /\s$/;
const ZERO_WIDTH_SPACE = '\u200b';

export function remarkTrailingWhitespace() {
    //
    // Append zero-width space to start of text nodes that have leading or trailing whitespace
    //    
    return ast => {
        visit(ast, REMOVE_TYPES, (node, index, parent) => {
            // Remove empty nodes
            if (!node.children.length) {
                parent.children.splice(index, 1);
                return SKIP;
            }

            // Make certain that empty text nodes are not empty, if they are add a zero-width space to make certain they don't end up as dividers
            // Ideally we just delete these.
            if(node.children.length === 1 && node.children[0].type === 'text' && node.children[0].value.trim() === ''){
                node.children[0].value = ZERO_WIDTH_SPACE;
                return CONTINUE;
            }

            const first = node.children[0];
            const last = node.children[node.children.length - 1];

            // Append zero-width space to start of text nodes that have leading whitespace
            
            if (first.type === 'text' && START_SPACE.test(first.value)) {
                first.value = `${ZERO_WIDTH_SPACE}${first.value.replace(ZERO_WIDTH_SPACE,'')}`;
            }

            // Append zero-width space to end of text nodes that have trailing whitespace
            
            if (last.type === 'text' && END_SPACE.test(last.value)) {
                last.value = `${last.value.replace('\u200b','')}${ZERO_WIDTH_SPACE}`;
            }
            return CONTINUE;
        });
    };
}
```

**Reply by wooorm** [reply]
Done in https://github.com/syntax-tree/mdast-util-to-markdown/releases/tag/2.1.1. `npm update` to get the fix! Work like this takes time. You can [sponsor us](https://unifiedjs.com/community/sponsor/) so that we can spend that time

**Reply by danburzo** [reply]
Thanks, @wooorm, works great!

**Comment by abelrasheed** [comment]
I initially tried to find a solution based on the solution by @danburzo but found out that there were some edge cases in which it also failed.
I think this was one of the cases :
`<em>italic<strong> bold</strong></em>` Here the leading space when taken out of the strong node will be added to emphasis node as a trailing space.

So I wrote my own recursion algorithm that bubbles the space up to the root node and also took care of a number of other possibilities where the translation can fail.
This particular solution was able to solve all the edge cases I was able to come up with.

If this is a valid solution I would like to contribute this code to the library where the translation is happening. Any help would be welcome


```javascript
const cleanUpSpaces = () => {
  return /** @param {import('hast').Nodes | import('mdast').TopLevelContent} htmlTree */ (htmlTree) => {
    /**
     * @param {import('hast').Parent} node
     * @param {import('hast').Parent | null} parent
     * @param {Number} index
     * @returns
     */
    const visitNode = (node, parent, index) => {
      // text nodes will not have any children property, so this will do an early return for all such tags.
      if (!node.children) {
        return;
      }

      /**
       * if strong, del or em tags doesn't have any children(children array is present,
       * but is an empty array),remove them as it can cause the translation to break.
       * eg: <p><strong></strong></p> -----> <p></p>
       */
      if (node.children.length === 0) {
        parent?.children.splice(index, 1);
        return;
      }

      /**
       * Traversing in reverse order because the visitNode function will change the array in which the node is present,
       * and in the case of leading spaces, the space gets extracted and inserted at the index of the node and the node gets displaced
       * into the next place, so in the case of forward iteration, this node is again processed.
       *
       * Also in the case when one node is deleted from the array, the next node will take it's place, and in the case
       * of forward iteration, this sibling node can get omitted.
       */

      for (let i = node.children.length - 1; i >= 0; i--) {
        const child = node.children[i];
        visitNode(child, node, i);
      }

      if (node.type !== 'strong' && node.type !== 'emphasis' && node.type !== 'delete') {
        return;
      }

      const firstChild = node.children[0];
      const lastChild = node.children[node.children.length - 1];

      if (firstChild.type === 'text') {
        /**
         * Looking for leading spaces:
         * <p><strong><text> Hello</text></strong></p> -----> <p><text> </text><strong><text>Hello</text></strong></p>
         */
        const [, leadingSpaces, textValue] = firstChild.value.match(/^(\s+)(.*?)$/) || [];
        if (leadingSpaces && leadingSpaces.length > 0) {
          firstChild.value = textValue;
          parent?.children.splice(index, 0, {
            type: 'text',
            value: leadingSpaces,
          });
          // index is incremented because the space was inserted in the place of the node, so the index of the node will change.
          index += 1;
        }
      }

      if (lastChild.type === 'text') {
        /**
         * Looking for trailing spaces:
         * <p><strong><text>Hello </text></strong></p> -----> <p><strong><text>Hello</text></strong><text> </text></p>
         */
        const [, textValue, trailingSpaces] = lastChild.value.match(/^(.*?)(\s+)$/) || [];

        if (trailingSpaces && trailingSpaces.length > 0) {
          lastChild.value = textValue;
          parent?.children.splice(index + 1, 0, {
            type: 'text',
            value: trailingSpaces,
          });
        }
      }

      /**
       * whenever a strong/del/em node has a single text node as a child and the value of the text node is an empty string,
       * remove the node.
       * eg: <p><strong><text></text></strong></p> ----> <p></p>; This can happen when there was a whitespace in the text
       *     node, which was identified as a leading space and was extracted outside.
       */
      if (firstChild === lastChild && firstChild.type === 'text' && firstChild.value === '') {
        parent?.children.splice(index, 1);
      }
      return;
    };
    visitNode(htmlTree, null, 0);
  };
};
```

**Reply by wooorm** [reply]
See https://github.com/orgs/syntax-tree/discussions/60#discussioncomment-2111096 above. In it, I explain in depth what the cause is and how it can be solved. It would need to be implemented in https://github.com/syntax-tree/mdast-util-to-markdown. I’d estimate 1, 2 days of work?

**Reply by abelrasheed** [reply]
Thanks for the reply, I will go through this thread !

**Reply by lakbychance** [reply]
Hey @abelrasheed / @wooorm , is there a plan of getting this `cleanupSpaces` solution as part of `toMarkdown` utility ? We are having same WYSIWYG problem as others pointed out. This utility seems to work on few cases we tried but wanted to know if it was not officially added due to some constraints ?

**Reply by wooorm** [reply]
No, see my earlier comment. It changes ASTs. It changes meaning. It changes content. That is what you can choose to do. It is not something `toMarkdown` does.
Above this thread I lay out a solution that would work. But. Something like it doesn’t exist yet. So it’s not simple: encoding the first or last character of a `text` node, when it is at an “emphasis break”, depending on what character is at the other end of that break.

I currently have a bit more time to look at open issues, [due to sponsorship](https://unifiedjs.com/community/sponsor/), so I may look at it soon.

**Reply by wooorm** [reply]
Done in https://github.com/syntax-tree/mdast-util-to-markdown/releases/tag/2.1.1. `npm update` to get the fix! Work like this takes time. You can [sponsor us](https://unifiedjs.com/community/sponsor/) so that we can spend that time

**Comment by kawanet** [comment]
Hi @wooorm , version 2.1.1 or commit https://github.com/syntax-tree/mdast-util-to-markdown/commit/97fb8181 has a breaking change on Markdown round trip.
It stringifies Markdown `foo***bar***buz` to `fo&#x6F;***bar***&#x62;uz`.
The issue https://github.com/syntax-tree/mdast-util-to-markdown/issues/66 filed.

**Comment by necauqua** [comment]
Oh my god, thank you so much @wooorm 💜
I checked out this issue out of nowhere today (I kind of abandoned a side-project incl. because of this and remembered about it) and you actually implemented your fix! So cool

**Reply by wooorm** [reply]
Thanks for the kind words Anton! :)
